foam.web.URLState: Application state tracking via URL hash value

mdittmer opened the pull request.

mdittmer committed:
foam.web.URLState: Application state tracking via URL hash value

mdittmer committed:
URLState: Fix up parser and hash observer pattern

mdittmer committed:
URLState: Clearer API names + two new methods for removing/clearing bindings; inner Window class for working with state independent of page URL hash

arobins commented on src/foam/web/URLState.js:
```diff
@@ -0,0 +1,236 @@
+        first = false;
+
+        value = this.serializer.stringify(bindingsMap[key].get());
+        hash += key + '=' + this.window.encodeURIComponent(value);
```
|key| should also be URI-encoded.

mdittmer replied:
Indeed! Fixed.

arobins commented on src/foam/web/URLState.js:
```diff
@@ -0,0 +1,236 @@
+      for ( var key in unboundMap ) {
+        if ( ! unboundMap.hasOwnProperty(key) ) continue;
+        value = this.serializer.stringify(unboundMap[key]);
+        hash += '&' + key + '=' + this.window.encodeURIComponent(value);
```
This breaks if there is nothing in bindingsMap.

mdittmer replied:
Indeed! Fixed.

arobins commented on src/foam/web/URLState.js:
```diff
@@ -0,0 +1,236 @@
+
+      var unboundMap = this.unboundMap_;
+      for ( var key in unboundMap ) {
+        if ( ! unboundMap.hasOwnProperty(key) ) continue;
```
This and the loop above are identical other than the input map. Maybe make a method for doing the conversion?

mdittmer replied:
In order for `first` handling to be general, it would need to be input/output, so the helper method would return two things (the new string and the `first` state).

IMHO it's simpler like this (though repetitive). [If there were 3](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) or more loops, I would tend to agree: a method would be better.

arobins replied:
Or just return a string, and let the caller deal with mashing them together. The logic for joining two potentially-empty strings is pretty trivial.

mdittmer replied:
On second thought, I can use an output function to avoid requiring two return values. Updated logic with helper function is in `DetachedURLState.js`

mdittmer committed:
URLState: Address some comments in #639 and refactor URLState:

- DetachedURLState contains core logic, but does not bind to
  window.location.hash

- URLState adds window.location.hash bindings

DetachedURLState is useful for generating URLs based on app state. URLState
is for "live updating" app URL.

arobins commented on src/foam/web/DetachedURLState.js:
```diff
@@ -0,0 +1,212 @@
+      function out(str) { hash += str; }
+
+      // Output bindings, then unbound values.
+      this.outputMapToHashStr_(
```
I feel like this would be easier to follow if the helper accepted a current string and appended key/value pairs to it. You can replace the first flag with an empty string check, and return the string at the end instead of the flag.

mdittmer replied:
Done. @arobins PTAL

mdittmer committed:
URLState: More imperative stateToHash_() helper

arobins reviewed (commented):
lgtm, but I'd suggest waiting for someone else from FOAM to review.

mdittmer committed:
URLState bug fix

Bound map contains Slots, unbound map contains values. Provide separate
getter functions and inject them in to hash string builder helper method.
